### PR TITLE
Use TapTweakHash::from_key_and_tweak() method in computing tweak for UntweakedPublicKey

### PR DESCRIPTION
**Pull request by nlanson**
Quick follow up PR to #691 using a method from #677.

### Changes
- Updated `UntweakedPublicKey::tap_tweak(...)` to use `TapTweakHash::from_key_and_tweak(...)`

### CODE REVIEW

**Review by dr-orlovsky — APPROVED** [review]
utACK 9e2bad304c36cd2d314db11b31d11be3b8f703c7

**Review by sanket1729 — APPROVED** [review]
ACK 9e2bad304c36cd2d314db11b31d11be3b8f703c7

File: src/util/schnorr.rs [hunk 1]
```diff
@@ -49,11 +49,7 @@ pub trait TapTweak {
 impl TapTweak for UntweakedPublicKey {
     fn tap_tweak<C: Verification>(&self, secp: Secp256k1<C>, merkle_root: Option<TapBranchHash>) -> TweakedPublicKey {
         // Compute the tweak
-        let mut engine = TapTweakHash::engine();
-        engine.input(&self.serialize());
-        merkle_root.map(|hash| engine.input(&hash));
-        let tweak_value: [u8; 32] = TapTweakHash::from_engine(engine).into_inner();
-            
+        let tweak_value = TapTweakHash::from_key_and_tweak(self.clone(), merkle_root).into_inner();
```
Comment (Kixunil) [inline src/util/schnorr.rs:52]: Isn't `clone()` expensive?

Comment (sanket1729) [reply src/util/schnorr.rs:52]: schnorr::PublicKey is Copy, I think it is the same.

Comment (Kixunil) [reply src/util/schnorr.rs:52]: In that case `.clone()` is redundant and I don't think it should be used.

Comment (nlanson) [reply src/util/schnorr.rs:52]: `TapTweakHash::from_key_and_tweak()` takes in `schnorrsig::PublicKey` where as the function it is being called in has `&UntweakedPublicKey`. 

I used the `.clone()` to get around the borrow checker. Should I instead pass in `*self` to the method or should I modify `TapTweakHash::from_key_and_tweak()` to take in `&schnorrsig::PublicKey` as a parameter?

Comment (Kixunil) [reply src/util/schnorr.rs:52]: I think taking `self` by value everywhere is the correct approach. The compiler should insert references for optimization if needed.

Comment (nlanson) [reply src/util/schnorr.rs]: Fixed in the latest commit and squashed.

**Comment by dr-orlovsky** [comment]
Needs rebase

**Comment by nlanson** [comment]
rebased

**Comment by nlanson** [comment]
I have tried to rebase but noticed that in #707 , the `merkle_root` parameter for `Address::p2tr` has been updated to take in `Option<&TapBranchHash>` but in `TapTweakHash::from_key_and_tweak()` it is `Option<TapBranchHash>` without the reference as it's parameter. 

Should `TapTweakHash::from_key_and_tweak()` be updated and changes reflected accross `taproot.rs` (a couple of spots need to be updated to use reference) or is this out of the scope of the PR?

**Comment by dr-orlovsky** [comment]
Generally I think we should take `Option<TapBranchHash>` and pass it by value. First, it's small enough and meets our agreement on pass-by-value we discussed recently; secondly it will be passed as `Some(hash)` and not variable to this API, so it does not make sense to force users to write `Some(&hash)` instead

**Comment by dr-orlovsky** [comment]
@nlanson: https://github.com/rust-bitcoin/rust-bitcoin/pull/695#pullrequestreview-815310088

**Comment by nlanson** [comment]
Updated the following methods to take `Option<TapBranchHash>` as value instead of reference:
 - `TapTweak::tap_tweak(...)`
 - `Address::p2tr(...)`

**Review by dr-orlovsky — APPROVED** [review]
utACK 5b21a9cb1fa818607d18a5654cd18a11efaaab4b

**Review by Kixunil — APPROVED** [review]

**Comment by Kixunil** [comment]
ACK 5b21a9cb1fa818607d18a5654cd18a11efaaab4b

(sorry for forgetting to add this into approve message - IDK if relevant)

**Pull request merged**